### PR TITLE
Verify the rollout status of the hosted router and registry

### PR DESCRIPTION
**Pull request by stevekuznetsov**
When deploying the hosted router and registry components, we need to
ensure that they correctly roll out. The previous checks were weak in
that they either simply waited for a set amount of time and/or did one
replica check. They would fail if the router or registry took longer to
deploy or if there were un-ready or failing replicas. The `oc rollout`
command group contains the `status` endpoint for internalizing all of
the logic for determining when a rollout has succeeded or failed, so
simply using this client call will ensure that the router and registry
correctly deploy.

Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

[test]
aos-ci-test

/cc @sdodson

### CODE REVIEW

**Comment by sdodson** [comment]
`oc rollout status` does what when it hasn't completed? And it's supported in which OCP versions?

**Comment by stevekuznetsov** [comment]
> `oc rollout status` does what when it hasn't completed?

Waits, determines if it's progressing or not. When it fails to progress and fails, the client call fails.

> And it's supported in which OCP versions?

3.5 and 3.6 for sure

**Comment by stevekuznetsov** [comment]
More specifically, 1.4+

```
$ tags=()
$ for commit in $( git log --pretty=%H -- contrib/completions/bash/oc ); do if git grep -q _oc_rollout_status "${commit}"; then tags+=( $( git tag --contains "${commit}" ) ); fi; done
$ echo "${tags[@]}" | sort | uniq
v1.4.0
v1.4.0-alpha.1
v1.4.0-rc1
v1.4.1
v1.5.0
v1.5.0-alpha.0
v1.5.0-alpha.1
v1.5.0-alpha.2
v1.5.0-alpha.3
v1.5.0-rc.0
v1.5.1
v3.6.0-alpha.0
v3.6.0-alpha.1
v3.6.0-alpha.2
```

**Comment by stevekuznetsov** [comment]
Have your newlines, `tox`

@sdodson can we get aos-ci on this and I think it's ready

**Comment by stevekuznetsov** [comment]
```

TASK [openshift_hosted : Ensure OpenShift router correctly rolls out] **********
task path: /usr/share/ansible/openshift-ansible/roles/openshift_hosted/tasks/router/router.yml:93
changed: [localhost] => (item={u'name': u'router', u'certificate': {u'keyfile': u'/etc/origin/master/openshift-router.key', u'certfile': u'/etc/origin/master/openshift-router.crt', u'cafile': u'/etc/origin/master/ca.crt'}, u'replicas': u'1', u'serviceaccount': u'router', u'namespace': u'default', u'stats_port': 1936, u'edits': [{u'action': u'put', u'value': 1, u'key': u'spec.strategy.rollingParams.intervalSeconds'}, {u'action': u'put', u'value': 1, u'key': u'spec.strategy.rollingParams.updatePeriodSeconds'}, {u'action': u'put', u'value': 21600, u'key': u'spec.strategy.activeDeadlineSeconds'}], u'images': u'openshift/origin-${component}:cc2ed8f', u'selector': u'region=infra', u'ports': [u'80:80', u'443:443']}) => {
    "rc": 0, 
    "stderr": [], 
    "stdout": [
        "Waiting for rollout to finish: 0 out of 1 new replicas have been updated...", 
        "Waiting for rollout to finish: 0 out of 1 new replicas have been updated...", 
        "Waiting for rollout to finish: 0 of 1 updated replicas are available...", 
        "Waiting for latest deployment config spec to be observed by the controller loop...", 
        "replication controller \"router-1\" successfully rolled out"
    ]
}
```

**Comment by stevekuznetsov** [comment]
jUnit reports:
 - [router](https://ci.openshift.redhat.com/jenkins/job/test_pull_request_openshift_ansible_extended_conformance_install_with_status_check/529/testReport/(root)/Create%20Hosted%20Resources/_localhost__openshift_hosted___Ensure_OpenShift_router_correctly_rolls_out) ran in 23s
 - [registry](https://ci.openshift.redhat.com/jenkins/job/test_pull_request_openshift_ansible_extended_conformance_install_with_status_check/529/testReport/(root)/Create%20Hosted%20Resources/_localhost__openshift_hosted___Ensure_OpenShift_registry_correctly_rolls_out) ran in 12s

**Comment by sdodson** [comment]
I'd expect up to a minute or more in environments that actually require images to be pulled. Especially for whichever is deployed first as pod, deployer, and the actual image need to be pulled.

**Comment by sdodson** [comment]
aos-ci-test

**Comment by stevekuznetsov** [comment]
> I'd expect up to a minute or more in environments that actually require images to be pulled. Especially for whichever is deployed first as pod, deployer, and the actual image need to be pulled.

I would too -- but the cluster isn't valid if this doesn't happen, so it's kind of a moot point. I am more worried that it would have made CI significantly slower.

**Comment by stevekuznetsov** [comment]
Job only failed on [checking other job statuses](https://ci.openshift.redhat.com/jenkins/job/test_pull_request_openshift_ansible/195/consoleFull#168877841158b6e51eb7608a5981914356)

**Comment by sdodson** [comment]
[merge]

**Comment by stevekuznetsov** [comment]
AOS CI decided not to work? WTF?

**Comment by stevekuznetsov** [comment]
aos-ci-test

**Comment by stevekuznetsov** [comment]
@sdodson looks like it's throwing a fit... no logs? Can we re-test?

**Comment by sdodson** [comment]
aos-ci-test

**Comment by kargakis** [comment]
I wouldn't use `rollout status` in a script/Ansible playbook before https://github.com/kubernetes/kubernetes/issues/40224 is fixed. @mfojtik was working at it but I am not sure of his status on it.

**Comment by stevekuznetsov** [comment]
@kargakis do we see that often? I've switched the Origin Aggregated Logging tests to use `oc rollout status` to verify their deployments, and [that job](https://ci.openshift.redhat.com/jenkins/job/test_pull_request_origin_aggregated_logging_ansible/) has more than 100 builds without anyone observing that issue.

**Comment by stevekuznetsov** [comment]
Or are you more concerned about customer environments rather than CI flake

**Comment by kargakis** [comment]
> Or are you more concerned about customer environments rather than CI flake

Exactly. kubernetes/kubernetes#40224 needs to be fixed for `rollout status` prime time.

**Comment by stevekuznetsov** [comment]
Hmmm -- can we add `failed_when: no` to this to make it best-effort? It will block until it either correctly determines that the rollout succeeded or fails on some bug (but we won't care).

In the interim do we have a better solution for verifying the rollout status? This hit some other people as well where it was as bad as the router image being missing and the installer reported 100% happy.

Should we just do the `oc rollout status` call, ignore the output, and the check the `.spec.replicas` versus the `.status` as a sanity check?

**Comment by kargakis** [comment]
`failed_when: no` followed by another task that verifies the status of the RC (check that openshift.io/deployment.phase == Complete).

**Comment by stevekuznetsov** [comment]
@kargakis why RC, not DC?

**Comment by stevekuznetsov** [comment]
Made the `rollout status` check best-effort and added a poll as a follow-up. @sdodson PTAL the ansible

File: roles/openshift_hosted/tasks/registry/registry.yml [hunk 1]
```diff
@@ -124,6 +124,33 @@
...
+- name: Sanity-check that the OpenShift registry rolled out correctly
+  command: |
+    oc get replicationcontroller {{ openshift_hosted_registry_name }}-{{ openshift_hosted_registry_latest_version.stdout }} \
+           --namespace {{ openshift_hosted_registry_namespace }} \
+           --config {{ openshift.common.config_base }}/master/admin.kubeconfig \
+           -o jsonpath='{ .metadata.annotations.openshift\.io/deployment\.phase }'
+  register: openshift_hosted_registry_rc_phase
+  until: "'Running' not in openshift_hosted_registry_rc_phase.stdout"
+  delay: 15
+  reties: 40
```
Comment (sdodson) [inline roles/openshift_hosted/tasks/registry/registry.yml]: retries

File: roles/openshift_hosted/tasks/router/router.yml [hunk 2]
```diff
@@ -89,18 +89,35 @@
...
+- name: Poll for OpenShift router deployment success
+  command: |
+    oc get replicationcontroller {{ item.0.name }}-{{ item.1.stdout }} \
+           --namespace {{ item.0.namespace }} \
+           --config {{ openshift.common.config_base }}/master/admin.kubeconfig \
+           -o jsonpath='{ .metadata.annotations.openshift\.io/deployment\.phase }'
+  register: openshift_hosted_router_rc_phase
+  until: "'Running' not in openshift_hosted_router_rc_phase.stdout"
+  delay: 15
+  reties: 40
```
Comment (sdodson) [inline roles/openshift_hosted/tasks/router/router.yml]: retries

File: roles/openshift_hosted/tasks/registry/registry.yml [hunk 3]
```diff
@@ -124,6 +124,33 @@
...
+  failed_when: "'Failed' in openshift_hosted_registry_rc_phase.stdout"
```
Comment (sdodson) [inline roles/openshift_hosted/tasks/registry/registry.yml]: Did you figure out if this fails early or does it hit max retries?

File: roles/openshift_hosted/tasks/registry/registry.yml [hunk 4]
```diff
@@ -124,6 +124,33 @@
     edits: "{{ openshift_hosted_registry_edits }}"
     force: "{{ True|bool in openshift_hosted_registry_force }}"
 
+- name: Ensure OpenShift registry correctly rolls out (best-effort today)
+  command: |
+    oc rollout status deploymentconfig {{ openshift_hosted_registry_name }} \
+                      --namespace {{ openshift_hosted_registry_namespace }} \
+                      --config {{ openshift.common.config_base }}/master/admin.kubeconfig
+  failed_when: no
```
Comment (sdodson) [inline roles/openshift_hosted/tasks/registry/registry.yml]: I think we tend to gravitate towards `false`

File: roles/openshift_hosted/tasks/registry/registry.yml (hunk 1 above)
Comment (stevekuznetsov) [inline roles/openshift_hosted/tasks/registry/registry.yml]: What if we wanted to tie it again, though?

File: roles/openshift_hosted/tasks/registry/registry.yml (hunk 3 above)
Comment (stevekuznetsov) [inline roles/openshift_hosted/tasks/registry/registry.yml]: It doesn't fail early, unfortunately. No way to do that it seems

File: roles/openshift_hosted/tasks/registry/registry.yml (hunk 4 above)
Comment (stevekuznetsov) [inline roles/openshift_hosted/tasks/registry/registry.yml]: ack

**Comment by stevekuznetsov** [comment]
@sdodson updated PTAL

**Comment by stevekuznetsov** [comment]
aos-ci-test

**Comment by stevekuznetsov** [comment]
PLEASE

**Comment by sdodson** [comment]
aos-ci-test

**Comment by sdodson** [comment]
[merge]

**Comment by sdodson** [comment]
[merge]

**Comment by stevekuznetsov** [comment]
@openshift-bot, the [last build](https://ci.openshift.redhat.com/jenkins/job/merge_pull_request_openshift_ansible/562/) failed from the following flakes:
 - [tox](https://ci.openshift.redhat.com/jenkins/job/test_pull_request_openshift_ansible_tox/473/consoleFull#84165338956bf4006e4b05b79524e5923) failed on timeouts:

```
--- FAIL: TestPackageUpdateRepoDisabled (3.11s)
	common.go:80: got exit code 2, want 0
	common.go:98: 
		$ (cd /data/src/github.com/openshift/openshift-ansible/test/integration/openshift_health_checker/preflight && ansible-playbook -i /dev/null playbooks/package_update_repo_disabled.yml)
		...
		PLAY RECAP *********************************************************************
		localhost                  : ok=1    changed=0    unreachable=0    failed=1   
		
		
		Failure summary:
		
		  1. Host:     localhost
		     Play:     Start container for specified test host
		     Task:     start container
		     Message:  Error connecting: Error while fetching server API version: Timeout value connect was Timeout(connect=60, read=60, total=None), but it must be an int, float or None.
```

**Comment by stevekuznetsov** [comment]
@sdodson can you re-tag for merge please

**Comment by sdodson** [comment]
[merge][severity: bug]

**Pull request merged**